pkg/scheduler/queue/batched.go

```diff
@@ -0,0 +1,233 @@
+	}
+	return placementKey.(PlacementKey), false
+}
+
+// Done marks a PlacementKey as done.
+func (bq *batchedProcessingPlacementSchedulingQueue) Done(placementKey PlacementKey) {
+	bq.active.Done(placementKey)
+	// It is OK for Done to be called on the batched queue even if the item has never been
+	// added to the batched queue before. In this case the call is simply a no-op.
+	bq.batched.Done(placementKey)
+}
+
+// Add adds a PlacementKey to the work queue for immediate processing.
+//
+// Note that this bypasses the rate limiter (if any).
+func (bq *batchedProcessingPlacementSchedulingQueue) Add(placementKey PlacementKey) {
+	bq.active.Add(placementKey)
+}
+
+// AddAfter adds a PlacementKey to the work queue after a set duration for immediate processing.
+//
+// Note that this bypasses the rate limiter (if any).
+func (bq *batchedProcessingPlacementSchedulingQueue) AddAfter(placementKey PlacementKey, duration time.Duration) {
+	bq.active.AddAfter(placementKey, duration)
+}
+
+// AddRateLimited adds a PlacementKey to the work queue after the rate limiter (if any)
+// says that it is OK, for immediate processing.
+func (bq *batchedProcessingPlacementSchedulingQueue) AddRateLimited(placementKey PlacementKey) {
+	bq.active.AddRateLimited(placementKey)
+}
+
+// Forget untracks a PlacementKey from rate limiter(s) (if any) set up with the queue.
+func (bq *batchedProcessingPlacementSchedulingQueue) Forget(placementKey PlacementKey) {
+	bq.active.Forget(placementKey)
+	// It is OK for Forget to be called on the batched queue even if the item has never been
+	// added to the batched queue before. In this case the call is simply a no-op.
+	bq.batched.Forget(placementKey)
+}
+
+// AddBatched tracks a PlacementKey and adds such keys in batch later to the work queue when appropriate.
+func (bq *batchedProcessingPlacementSchedulingQueue) AddBatched(placementKey PlacementKey) {
+	bq.batched.Add(placementKey)
+}
+
+// Run starts the scheduling queue.
+func (bq *batchedProcessingPlacementSchedulingQueue) Run() {
+	// Spin up a goroutine to move items periodically from the batched queue to the active queue.
+	go func() {
+		timer := time.NewTimer(time.Duration(bq.movePeriodSeconds) * time.Second)
+		for {
+			select {
+			case _, closed := <-bq.moveNow:
+				if closed && bq.batched.ShuttingDown() {
+					// The batched queue has been shut down, and the moveNow channel has been closed;
+					// now it is safe to assume that after moving all the items from the batched queue to the active queue
+					// this time, the batched queue will be drained.
+					bq.moveAllBatchedItemsToActiveQueue()
+					return
+				}
+
+				// The batched queue might still be running; move all items and re-enter the loop.
+				bq.moveAllBatchedItemsToActiveQueue()
+			case <-timer.C:
+				// The timer has fired; move all items.
+				bq.moveAllBatchedItemsToActiveQueue()
+			}
+
+			// Reset the timer for the next round.
+			timer.Reset(time.Duration(bq.movePeriodSeconds) * time.Second)
```

Comment:
why do we reset the timer for `case _, closed := <-bq.moveNow:` ?

Reply:
Hi Wei! The moving process is set to wait for 300 seconds after each moving attempt; I thought about using a tick-based mechanism (instead of waiting for 300 seconds after each move, do a move every 300 seconds regardless) but it has its own concerns.

Reply:
(missing ticks, two consecutive moves if we are unlucky, etc.)